library/src/main/java/com/novoda/downloadmanager/DownloadBatch.java

```diff
@@ -49,73 +53,144 @@ void setCallback(DownloadBatchStatusCallback callback) {
     }
 
     void download() {
-        DownloadBatchStatus.Status status = downloadBatchStatus.status();
+        String rawBatchId = downloadBatchStatus.getDownloadBatchId().rawId();
+        Log.v("start download " + rawBatchId + ", status: " + downloadBatchStatus.status());
 
-        if (status == DELETED) {
-            deleteBatchIfNeeded();
+        if (shouldAbortStartingBatch(connectionChecker, callback, downloadBatchStatus, downloadsBatchPersistence)) {
+            Log.v("abort starting download " + rawBatchId);
             return;
         }
 
-        if (status == PAUSED) {
-            return;
-        }
-
-        if (connectionNotAllowedForDownload(status)) {
-            processNetworkError();
-            return;
-        }
-
-        if (status != DOWNLOADED) {
-            downloadBatchStatus.markAsDownloading(downloadsBatchPersistence);
-            notifyCallback(downloadBatchStatus);
-        }
+        markAsDownloadingIfNeeded(downloadBatchStatus, downloadsBatchPersistence, callback);
 
         if (totalBatchSizeBytes == 0) {
-            totalBatchSizeBytes = getTotalSize(downloadFiles);
+            totalBatchSizeBytes = getTotalSize(downloadFiles, downloadBatchStatus);
         }
 
-        if (totalBatchSizeBytes <= ZERO_BYTES) {
-            deleteBatchIfNeeded();
-            processNetworkError();
+        if (shouldAbortAfterGettingTotalBatchSize(downloadBatchStatus, downloadsBatchPersistence, callback, totalBatchSizeBytes)) {
```

Comment:
This is exactly what I envisioned when all of these issues started cropping up. Perform an action, check that we can continue 💯